### PR TITLE
dist/tools/ci: add can_fast_ci_run.py

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

This PR add a simply script to check if a PR could be tested with a fast CI run that only builds a subset of apps and/or all apps for a subset of boards.

Check the [provided documentation of the tool](https://github.com/maribu/RIOT/tree/dist/tools/ci/dist/tools/ci#checking-if-fast-ci-runs-are-sufficient) for more details.

### Testing procedure

This PR contains a commit that contains a bunch of bogus changes that do not justify full rebuild. Run the script as specified and it should give you a list of build targets as JSON.

### Issues/PRs references

None

### CODE REVIEW

**Comment by maribu** [comment]
Note: Most of the script logic is for classifying the changed files into different areas, so that simple and readable heuristics can be written. These are the actual heuristics:

https://github.com/maribu/RIOT/blob/c2eae6407495b6b668205d5624aabaf6da74bf09/dist/tools/ci/can_fast_ci_run.py#L194-L217

I hope that eventually the list of changed modules / packages could be used to deduce which applications are impacted by this. This is why I decided to spent the effort to actually figure out which modules / packages are modified by the changes. I think that because KConfig is machine readable, this would enable efficiently parsing the dependencies and enable such feature.

File: bootloaders/riotboot_serial/main.c [hunk 1]
```diff
@@ -70,7 +70,7 @@ static void _boot_default(int slot)
     }
 }
 
-void kernel_init(void)
+NORETURN void kernel_init(void)
```
Comment (maribu) [inline bootloaders/riotboot_serial/main.c:73]: This change looks actually legitimate to me :-D

File: dist/tools/ci/can_fast_ci_run.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+
+EXCEPTION_MODULES = {"boards/common/nrf52"}
+
+
+def print_change_set_section(name, contents):
+    """
+    Print the given change set section human reable
+    """
+    if len(contents) > 0:
+        print(name, file=sys.stderr)
```
Comment (aabadie) [inline dist/tools/ci/can_fast_ci_run.py:43]: Everywhere print is printing to `sys.stderr`, this looks like a perfect use case for `functools.partial`:

```python
from functools import partial

print_err = partial(print, file=sys.stderr)

print_err("Should be print only to stderr")
```

File: dist/tools/ci/can_fast_ci_run.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+REGEX_APP = re.compile(r"^(bootloaders|examples|fuzzing|tests)\/")
+
+EXCEPTION_MODULES = {"boards/common/nrf52"}
+
+
+def print_change_set_section(name, contents):
+    """
+    Print the given change set section human reable
+    """
+    if len(contents) > 0:
```
Comment (aabadie) [inline dist/tools/ci/can_fast_ci_run.py:42]: You could check for emptyness of contents and return immediately:

```python
if not contents:
    return
```

File: dist/tools/ci/can_fast_ci_run.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+        print(name, file=sys.stderr)
+        print("=" * len(name), file=sys.stderr)
+        print("", file=sys.stderr)
+        for category in sorted(contents):
+            print(category, file=sys.stderr)
+            print("-" * len(category), file=sys.stderr)
+            print("", file=sys.stderr)
+            for file in sorted(contents[category]):
+                print(f"- {file}", file=sys.stderr)
```
Comment (aabadie) [inline dist/tools/ci/can_fast_ci_run.py:51]: f-strings unfortunately won't work on RPI workers (still using Python 3.5, too bad :( )

File: dist/tools/ci/can_fast_ci_run.py [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+            print("", file=sys.stderr)
+
+
+class ChangeSet:
+    """
+    Representation of the modules affected by a change set
+    """
+    def __init__(self, riotbase=None):
+        if riotbase is None:
+            riotbase = os.getcwd()
```
Comment (aabadie) [inline dist/tools/ci/can_fast_ci_run.py:61]:
```suggestion
    def __init__(self, riotbase=os.getcwd()):
```

File: dist/tools/ci/can_fast_ci_run.py [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+        self.apps = {}
+        self.boards = {}
+        self.modules = {}
+        self.other = {}
+        self.pkgs = {}
+        self.__riotbase = os.path.normpath(riotbase)
```
Comment (aabadie) [inline dist/tools/ci/can_fast_ci_run.py:67]: In Pep8, the convention for internal use attribute is to use a single leading underscore:

```
_single_leading_underscore: weak "internal use" indicator. E.g. from M import * does not import objects whose names start with an underscore.
```

File: dist/tools/ci/can_fast_ci_run.py [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+
+            raise Exception("Failed to parse \"{}\"".format(line))
+
+    return change_set
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="Check if a fast CI run is possible and which "
+                                     + "boards / applications to build")
+    parser.add_argument("--explain", default=False, action="store_const", const=True,
```
Comment (aabadie) [inline dist/tools/ci/can_fast_ci_run.py:176]: You could do this with the same result (here and below):

```suggestion
    parser.add_argument("--explain", action="store_true",
```

**Review by aabadie — COMMENTED** [review]
Nice one, I like the idea. I put some minor comments to improve the Python code a bit. Haven't checked in details how it works...

File: dist/tools/ci/can_fast_ci_run.py (hunk 5 above)
Comment (maribu) [reply dist/tools/ci/can_fast_ci_run.py:61]: This indeed doesn't need late-bound defaults. But it makes it easier to work with argparse (which just can use `None` as default, knowing that the `None` is later turned into something useful).

Comment (maribu) [reply dist/tools/ci/can_fast_ci_run.py:61]: Never mind, just using `default=os.getcwd()` there works fine and is more obvious.

Comment (aabadie) [reply dist/tools/ci/can_fast_ci_run.py]: Ah yes, so why not set `os.getcwd()` as default in argparse directly ?

**Comment by kaspar030** [comment]
I was somewhat hoping for a split between the tool and it's data, e.g., to have some json or yml file that describes the matching rules for a set. :/

**Comment by maribu** [comment]
> I was somewhat hoping for a split between the tool and it's data, e.g., to have some json or yml file that describes the matching rules for a set. :/

tl;dr: I don't think that we will need many more rules. With so little rules, I don't think it is worth the effort.

I can see the motivation for this. But I don't think that this is worth it here. Note that such a yaml file would need to have some defined syntax and associated semantics. I think it is fair to consider such a rule description format as a domain specific language. We will have to come up with a DSL that is as simple, easy to read, and compact as possible, while still being capable enough to allow expressing all rules we want to write. Such a DSL will have the advantage of being more compact and simpler than a full blown programming language for expressing rules.

If we write the rules in Python, we know that the language to express the rules is capable enough (due to being Turing complete), relatively compact and easy to read (as Python is famous for that), and most people understand Python syntax. So IMO a custom rule format will only play out it's advantages over coding the rules in Python, if there is a certain amount of rule.

I don't think that there will be much more rules to add. I hope that Kconfig will allow efficiently resolving and representing the dependencies, so that for any given module (or package), it can yield a list of applications and boards affected by changes in that module. This is the only thing currently missing that comes to my mind.

**Comment by kaspar030** [comment]
> tl;dr: I don't think that we will need many more rules. With so little rules, I don't think it is worth the effort.

OK. I'll focus on how to integrate this.

(TBH my first impression was that this could need some seperation of data and logic, but what couldn't...)

**Comment by maribu** [comment]
> (TBH my first impression was that this could need some seperation of data and logic, but what couldn't...)

I mean, I still can try and see where this gets us. Sometimes it really better to just see (prototypes of) both implementations, rather than wild-guessing how implementations would look like.

**Comment by aabadie** [comment]
> to have some json or yml file that describes the matching rules for a set

Even if I tend to prefer yml, be aware that using this format will require an external Python package ([PyYAML](https://pypi.org/project/PyYAML/)), so an external dependency, whereas json is available in the Python standard library.
Maybe this is not issue if the tool if meant to be used on the CI ?

**Comment by kaspar030** [comment]
> I mean, I still can try and see where this gets us. Sometimes it really better to just see (prototypes of) both implementations, rather than wild-guessing how implementations would look like.

Maybe if the classifiers can be represented as python dict, at the top of the file, that would be a first step.
We can then later load those from sth else (json, yml), if needed.

**Comment by maribu** [comment]
Ah, OK. I assumed you were referring to the decision rules, not the classifiers. I'll try to come up with something.

**Comment by kaspar030** [comment]
> Even if I tend to prefer yml, be aware that using this format will require an external Python package ([PyYAML](https://pypi.org/project/PyYAML/)), so an external dependency, whereas json is available in the Python standard library.
> Maybe this is not issue if the tool if meant to be used on the CI ?

no issue, we'd just add PyYAML to riotdocker / static tests.

**Comment by kaspar030** [comment]
I've been working on integrating this into .murdock. I need minimal additions to this script, but can put them into the integration PR.
If noone sees any blockers, I'd like to merge this.

**Review by kaspar030 — APPROVED** [review]
I did some testing, it seems to work as advertized.

ACK.

**Comment by kaspar030** [comment]
@maribu, please cleanup and squash!

**Comment by maribu** [comment]
Done.

Sorry that I didn't found the time so far to split out the classification patterns from the script. But I guess getting this in as is will ease collaboration on getting this into shape.
